drivers: flash: fix hyperflash write

jeppenodgaard opened the pull request:
This PR fixes hyperflash write issues.
It implements the functionality from mcux-sdk-examples on board evkbimxrt1050:

- [flexspi_hyper_flash_ops.c](https://github.com/nxp-mcuxpresso/mcux-sdk-examples/blob/main/evkbimxrt1050/driver_examples/flexspi/hyper_flash/polling_transfer/flexspi_hyper_flash_ops.c) contains the clock logic.
- [flexspi_hyper_flash_polling_transfer.c](https://github.com/nxp-mcuxpresso/mcux-sdk-examples/blob/main/evkbimxrt1050/driver_examples/flexspi/hyper_flash/polling_transfer/flexspi_hyper_flash_polling_transfer.c) contains the flash_flexspi_hyperflash_lut array.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/53855

Commit:
drivers: memc: add update clock function

Add a function to update the flexspi bus clock. This is
needed when write operations are done to the hyperflash.

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

Commit:
drivers: flash: fix hyperflash write operations

Hyperflash write operations resulted in invalid writes. This commit
fixes the issue by temporarily lowering the clock during writes.
This aligns with the mcux-sdk-examples.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/53855

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

pdgendt commented on drivers/memc/memc_mcux_flexspi.c:
```diff
@@ -72,6 +72,32 @@ int memc_flexspi_update_lut(const struct device *dev, uint32_t index,
+	}
+
+	FLEXSPI_Enable(data->base, false);
+	CLOCK_DisableClock(kCLOCK_FlexSpi);
```
You might need to include the clock symbols in this file

```c
#include <fsl_clock.h>
```

jeppenodgaard replied:
I need to rework the code. This will not work for all platforms..
I am trying to figure out a way but I am open for any input!

danieldegrasse replied:
One way to solve this may be adding an SOC specific function to support changing the FlexSPI clock dividers. This function could then be implemented in the soc_rt10xx.c file, and be supported for other SOCs using this hyperflash in the future if required.

jeppenodgaard replied:
Thanks! I reworked with your suggestions.

Devices that uses hyperflash but do not have an `imxrt_flexspi_clock_set_div` implementation will not build as far as I can tell.

Commit:
soc: rt10xx: add flexspi clock functions

Add SOC specific function to set the flexspi clock divider and get the
flexspi clock frequency.

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

pdgendt commented on drivers/memc/memc_mcux_flexspi.c:
```diff
@@ -72,6 +73,25 @@ int memc_flexspi_update_lut(const struct device *dev, uint32_t index,
+int memc_flexspi_update_clock(const struct device *dev,
+		enum memc_flexspi_clock_t clock)
+{
+#if CONFIG_FLASH_MCUX_FLEXSPI_HYPERFLASH
```
I don't think this symbol would be the preferred one to check?

jeppenodgaard replied:
`CONFIG_DT_HAS_NXP_IMX_FLEXSPI_ENABLED`? Or maybe `CONFIG_HAS_MCUX_FLEXSPI` or `CONFIG_CODE_FLEXSPI`?

pdgendt replied:
Maybe @danieldegrasse can suggest a beter way?

IMO, ideally the clocks would come from device tree bindings and get abstracted without the need for special casing in SoC code. But maybe that would take us too far in this PR.

danieldegrasse replied:
I agree- clocks from devicetree bindings is indeed the right way to do this. I consider the SOC function a short term implementation, since device tree based clocking is probably not something we want to hold this PR up with

Another option here would be to only call the soc clock reconfiguration function if the `SOC_SERIES_IMX_RT10XX` KConfig is set, because this is the only SOC implementing this support. I would also return `ENOTSUP` if the `imxrt_flexspi_clock_set_div` is not called, since the clock is not actually reconfigured.

One other nit: I'd prefer this function to not include the `imxrt` prefix- it is possible that other families could implement FlexSPI in the future, and need this function.

danieldegrasse commented:
@jeppenodgaard how are you testing this fix? When I run `samples/drivers/flash_shell`, I still see the following behavior:

```
uart:~$ flash_sample write_pattern 0x340020 100
Reading back written bytes:
00 01 02 03 | 04 05 06 07 | 08 09 0a 0b | 0c 0d 0e 0f
00 01 02 03 | 04 05 06 07 | 08 09 0a 0b | 0c 0d 0e 0f
00 00 00 00 | 00 00 00 00 | 00 00 00 00 | 00 00 00 00
00 01 02 03 | 04 05 06 07 | 08 09 0a 0b | 0c 0d 0e 0f
00 01 02 03 | 04 05 06 07 | 08 09 0a 0b | 0c 0d 0e 0f
00 01 02 03 | 04 05 06 07 | 08 09 0a 0b | 0c 0d 0e 0f
00 01 02 03 |
uart:~$
```

Note the block of cleared bytes at offset `0x340020`.

jeppenodgaard commented:
> @jeppenodgaard how are you testing this fix? When I run `samples/drivers/flash_shell`, I still see the following behavior:
> 
> ```
> uart:~$ flash_sample write_pattern 0x340020 100
> Reading back written bytes:
> 00 01 02 03 | 04 05 06 07 | 08 09 0a 0b | 0c 0d 0e 0f
> 00 01 02 03 | 04 05 06 07 | 08 09 0a 0b | 0c 0d 0e 0f
> 00 00 00 00 | 00 00 00 00 | 00 00 00 00 | 00 00 00 00
> 00 01 02 03 | 04 05 06 07 | 08 09 0a 0b | 0c 0d 0e 0f
> 00 01 02 03 | 04 05 06 07 | 08 09 0a 0b | 0c 0d 0e 0f
> 00 01 02 03 | 04 05 06 07 | 08 09 0a 0b | 0c 0d 0e 0f
> 00 01 02 03 |
> uart:~$
> ```
> 
> Note the block of cleared bytes at offset `0x340020`.

@danieldegrasse I tested with the dfu sample and transferred an image. I have not tested the newest fix because I did not bring the hardware home. I wanted to mention that I would test this when I get access to the hardware tomorrow but forgot, sorry.

The timezone difference is slightly annoying.

I will fix it tomorrow.

Despite it not working now, is the implementing acceptable?

jeppenodgaard commented:
@danieldegrasse the test you ran did not change the clock since I used `SOC_SERIES_IMX_RT10XX` without the CONFIG_ prefix. This is now fixed.

The soc files have been reworked slightly so only the flexspi clock divider function is relocated to ITCM. Tests with `samples/drivers/flash_shell/` on `mimxrt1050_evk` looks good to me.

I changed the `write_pattern` function so it compares write and read and then tested with `flash_sample write_pattern 0x340000 0x2000` which does not give any error.

I do not have access to other rt10xx or rt11xx boards so I cannot test that hardware.

Test output excluding the length 0x2000 test:
```
uart:~$ flash_sample page_read 13 100
00 00 01 00 | 01 02 03 04 | 00 01 02 03 | 04 05 06 07
08 09 0a 0b | 0c 0d 0e 0f | 10 11 12 13 | 14 15 16 17
18 19 1a 1b | 1c 1d 1e 1f | 20 21 22 23 | 24 25 26 27
28 29 2a 2b | 2c 2d 2e 2f | 30 31 32 33 | 34 35 36 37
38 39 3a 3b | 3c 3d 3e 3f | 40 41 42 43 | 44 45 46 47
48 49 4a 4b | 4c 4d 4e 4f | 50 51 52 53 | 54 55 56 57
58 59 5a 5b |
uart:~$ flash_sample page_erase 13 1
Erasing page 13 (start offset 0x340000, size 0x40000)
uart:~$ flash_sample page_read 13 100
ff ff ff ff | ff ff ff ff | ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff | ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff | ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff | ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff | ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff | ff ff ff ff | ff ff ff ff
ff ff ff ff |
uart:~$ flash_sample write_pattern 0x340000 0x1
Reading back written bytes:
00
uart:~$ flash_sample write_pattern 0x340004 0x2
Reading back written bytes:
00 01
uart:~$ flash_sample write_pattern 0x340008 0x3
Reading back written bytes:
00 01 02
uart:~$ flash_sample write_pattern 0x34000c 0x4
Reading back written bytes:
00 01 02 03 |
uart:~$ flash_sample page_read 13 0x20
00 ff ff ff | 00 01 ff ff | 00 01 02 ff | 00 01 02 03
ff ff ff ff | ff ff ff ff | ff ff ff ff | ff ff ff ff
uart:~$ flash_sample write_pattern 0x340010 100
Reading back written bytes:
00 01 02 03 | 04 05 06 07 | 08 09 0a 0b | 0c 0d 0e 0f
10 11 12 13 | 14 15 16 17 | 18 19 1a 1b | 1c 1d 1e 1f
20 21 22 23 | 24 25 26 27 | 28 29 2a 2b | 2c 2d 2e 2f
30 31 32 33 | 34 35 36 37 | 38 39 3a 3b | 3c 3d 3e 3f
40 41 42 43 | 44 45 46 47 | 48 49 4a 4b | 4c 4d 4e 4f
50 51 52 53 | 54 55 56 57 | 58 59 5a 5b | 5c 5d 5e 5f
60 61 62 63 |
uart:~$ flash_sample page_erase 13 1
Erasing page 13 (start offset 0x340000, size 0x40000)
```

danieldegrasse reviewed (commented):
Hi @jeppenodgaard, I have tested this revision, and can confirm that it works with the `flash_shell` sample. However, the `flash_shell` sample also seems to work on the RT1050 EVK on the latest revision of `main`. Have you tested that sample there? In #53855 it seems you were using the `flash` commands, not `flash_shell`.

I've attempted to verify MCUBoot support with this PR, and it has thus far not worked for me. I'm using the following applications:

1. MCUBoot
2. SMR server sample with -DOVERLAY_CONFIG='overlay-shell.conf'
3. Hello world (as the upgrade application)

I'm building and flashing with the following commands:
```
> west build -p always -b mimxrt1050_evk -d rt1050_bootloader/ bootloader/mcuboot/boot/zephyr/
> west build -p always -d rt1050_dfu -b mimxrt1050_evk zephyr/samples/subsys/mgmt/mcumgr/smp_svr/ -DOVERLAY_CONFIG='overlay-shell.conf'
> west build -p always -b mimxrt1050_evk -d rt1050_update zephyr/samples/hello_world/ -DCONFIG_BOOTLOADER_MCUBOOT=y

> west flash -d rt1050_bootloader/
> west sign -d rt1050_dfu -t imgtool -- -k bootloader/mcuboot/root-rsa-2048.pem
> west sign -d rt1050_update/ -t imgtool -- -k bootloader/mcuboot/root-rsa-2048.pem

> west flash -d rt1050_dfu/ -f rt1050_dfu/zephyr/zephyr.signed.bin
```

I then upload the image with the following commands:
```
> mcumgr -c acm0 image upload rt1050_update/zephyr/zephyr.signed
.bin
> mcumgr -c acm0 image list
Images:
 image=0 slot=0
    version: 0.0.0
    bootable: true
    flags: active confirmed
    hash: b79db39a57757fe2dcc9df4511e8bdd31fbcabba1b0d73c1cc4c988ac30c67f7
 image=0 slot=1
    version: 0.0.0
    bootable: true
    flags:
    hash: 7f12b01810556d447a8477bb5903ad89b0396cf3986f6fabbe247c5af6bcdf5f
Split status: N/A (0)
> mcumgr -c acm0 image test 7f12b01810556d447a8477bb5903ad89b0396cf3986f6fabbe247c5af6bcdf5f
Images:
 image=0 slot=0
    version: 0.0.0
    bootable: true
    flags: active confirmed
    hash: b79db39a57757fe2dcc9df4511e8bdd31fbcabba1b0d73c1cc4c988ac30c67f7
 image=0 slot=1
    version: 0.0.0
    bootable: true
    flags: pending
    hash: 7f12b01810556d447a8477bb5903ad89b0396cf3986f6fabbe247c5af6bcdf5f
```

On reset, the bootloader enters a bootloop, and swap does not work correctly. I assume there is some issue with the flash write algorithm, but the specific issue is not clear.

jeppenodgaard commented:
@danieldegrasse flash_sample write_pattern works on main (`10130d8e6d`) flash write does not.
main:
```
uart:~$ flash read 0x340040 0x30
00340040: ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff |........ ........|
00340050: ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff |........ ........|
00340060: ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff |........ ........|
uart:~$ flash write 0x340040 0x12 0x34 0x56 0x78
Write OK.
Verification ERROR!
uart:~$ flash write 0x340050 0x12 0x34 0x56 0x78
Write OK.
Verification ERROR!
uart:~$ flash read 0x340040 0x30
00340040: 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00 |........ ........|
00340050: 00 00 20 00 00 00 00 00  00 00 00 00 00 00 00 00 |.. ..... ........|
00340060: ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff |........ ........|
```
bugfix/mimxrt1050_hyperflash_write:
```
uart:~$ flash read 0x340040 0x30
00340040: ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff |........ ........|
00340050: ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff |........ ........|
00340060: ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff |........ ........|
uart:~$ flash write 0x340040 0x12 0x34 0x56 0x78
Write OK.
Verified.
uart:~$ flash write 0x340050 0x12345678
Write OK.
Verified.
uart:~$ flash read 0x340040 0x30
00340040: 12 00 00 00 34 00 00 00  56 00 00 00 78 00 00 00 |....4... V...x...|
00340050: 78 56 34 12 ff ff ff ff  ff ff ff ff ff ff ff ff |xV4..... ........|
00340060: ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff |........ ........|
```

I can reproduce the boot loop. However it succeeds after 5 loops (only tested this once):
[dfu.txt](https://github.com/zephyrproject-rtos/zephyr/files/10477741/dfu.txt)

jeppenodgaard commented on soc/arm/nxp_imx/rt/CMakeLists.txt:
```diff
@@ -29,6 +29,11 @@ if(CONFIG_PM)
   zephyr_sources_ifdef(CONFIG_SOC_SERIES_IMX_RT11XX power_rt11xx.c)
 endif()
 
+if (CONFIG_FLASH_MCUX_FLEXSPI_XIP AND CONFIG_SOC_SERIES_IMX_RT10XX)
```
Fixed.

jeppenodgaard commented on drivers/flash/flash_mcux_flexspi_hyperflash.c:
```diff
@@ -96,7 +96,7 @@ static const uint32_t flash_flexspi_hyperflash_lut[CUSTOM_LUT_LENGTH] = {
 	/* Write Enable */
 	[4 * WRITE_ENABLE] =
-		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_DDR, kFLEXSPI_8PAD, 0x20,
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_DDR, kFLEXSPI_8PAD, 0x00,
```
Fixed. The original array matches the sdk example.

danieldegrasse commented:
> While debugging I set MCUBOOT_LOG_LEVEL_DBG=y and added BOOT_LOG_DBG("%s:%u", __FILE__, __LINE__) before and after each function call in swap_scratch.c:boot_swap_sectors() and the issue no longer occurs for me.
>
> If MCUboot is built default MCUBOOT_LOG_LEVEL_LVL but NO_OPTIMIZATIONS=y is set it does not boot loop either.
>
> My guess is that in both cases there is a delay between the write/erase calls which fixes the issue. Got any ideas?


One thing to consider is that I notice a difference between your implementation of the clock update sequence versus the SDK's: the SDK calls FLEXSPI_UpdateDllValue() after reconfiguring the clock speed (https://github.com/nxp-mcuxpresso/mcux-sdk-examples/blob/main/evkbimxrt1050/driver_examples/flexspi/hyper_flash/polling_transfer/flexspi_hyper_flash_ops.c#L336). This function does include a 100 NOP delay here: https://github.com/zephyrproject-rtos/hal_nxp/blob/master/mcux/mcux-sdk/drivers/flexspi/fsl_flexspi.c#L466, which could explain why placing a delay before writes from MCUBoot fixes the swap support.

jeppenodgaard commented:
@danieldegrasse I looked at that and I have committed what I've done in the newest commit. The system is more unstable with those changes. I am not able to transfer an image successfully. Maybe something is not aligned with the example.

danieldegrasse commented:
> @danieldegrasse I looked at that and I have committed what I've done in the newest commit. The system is more unstable with those changes. I am not able to transfer an image successfully. Maybe something is not aligned with the example.

The example does not seem to demonstrate this, but typically changing the FlexSPI clock divider while running in XIP mode can cause issues. While the functions you have added to update clock frequencies are located in ITCM, the HAL `CLOCK_*` API functions likely aren't. Take a look at the implementation here for changing FlexSPI clock sources to support low power mode: https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/nxp_imx/rt/lpm_rt1064.c#L261

Something like this might help resolve the issue (although I'm unsure if it will given that these functions don't appear to be relocated in the MCUX SDK sample)

jeppenodgaard commented:
> > @danieldegrasse I looked at that and I have committed what I've done in the newest commit. The system is more unstable with those changes. I am not able to transfer an image successfully. Maybe something is not aligned with the example.
> 
> The example does not seem to demonstrate this, but typically changing the FlexSPI clock divider while running in XIP mode can cause issues. While the functions you have added to update clock frequencies are located in ITCM, the HAL `CLOCK_*` API functions likely aren't. Take a look at the implementation here for changing FlexSPI clock sources to support low power mode: https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/nxp_imx/rt/lpm_rt1064.c#L261
> 
> Something like this might help resolve the issue (although I'm unsure if it will given that these functions don't appear to be relocated in the MCUX SDK sample)

I will wait with PM for now.

The CLOCK_* functions are `static inline` header functions, so I guess they are relocated along with `flexspi_rt10xx.c`.
Anyway I tried relocating the header and source but it did not improve.

jeppenodgaard commented:
Without changing the DLL the app works. If I change the DLL crashes most of the time when I transfer the image via USB DFU. Maybe the USB uses the same DLL and therefore something goes wrong during the transfer.

I have not managed to make the boot loader work consistently yet and I am a bit stuck.
I tried adding the 100 NOPs after disabling/changing/enabling the divider.

@danieldegrasse do you have any ideas?

danieldegrasse commented:
@jeppenodgaard I believe your issue still has to do with code relocation. If you look at the linker file for the MCUX SDK sample, you can see that the SDK relocates the flexSPI and clock drivers to TCM: https://github.com/nxp-mcuxpresso/mcux-sdk-examples/blob/main/evkbimxrt1050/driver_examples/flexspi/hyper_flash/polling_transfer/armgcc/MIMXRT1052xxxxx_flexspi_nor.ld#L89. This is important because the function `CLOCK_GetFreq` call made in `flexspi_clock_get_freq` is *not inline*, and will be in flash unless relocated. I've created a commit that moves the HAL clock driver to ITCM when the flexSPI driver is enabled, which you can see here: https://github.com/nxp-zephyr/hal_nxp/commit/8f5c9b7c9821da65b44860708bea8d0b3633c2b3

With this commit I have been able to get MCUBoot working, both with `tests/boot/test_mcuboot` and the DFU upgrade sample. I've also tested the flash shell (*not the flash_sample shell*) with this PR and `main`, and can confirm the `flash write` command exposes issues with flexSPI driver unless the clocks are reconfigured. It seems that the way the `flash_sample` shell issues writes does not expose this issue.

Given that we are relocating the clock driver to RAM (which will add the `KEEP` flag to all relocated functions), and including new clock configuration code at the SOC level, I'd like to guard this behavior with a Kconfig. Something like `CONFIG_FLASH_MCUX_FLEXSPI_CLOCK_CONFIG`, which when enabled will:
- include SOC-specific clocking functions
- relocate `fsl_clock.c` to RAM in NXP's HAL
- compile in the `memc_flexspi_update_clock` function

jeppenodgaard commented:
It works sometimes but it is still unstable for me.

The ITCM size of the images increased so I the fsl_clock functions should be relocated:

| image | Flash before | Flash after | ITCM before | ITCM after |
| - | - | - | - | - |
| boot loader | 53758 B | 56174 B | 4504 B | 8120 B |
| dfu sample | 89836 B | 91876 B | 3656 B | 6840 B |

It crashes during `Download        [========                 ]  32%         6272 bytesdfu-util: Error during download get_status`. The Address is not consistent:

```
*** Booting Zephyr OS build zephyr-v3.2.0-4025-g0549333eb0d1 ***
[00:00:00.000,000] <inf> main: This device supports USB DFU class.

[00:00:06.972,000] <err> os: ***** USAGE FAULT *****
[00:00:06.973,000] <err> os:   Unaligned memory access
[00:00:06.973,000] <err> os: r0/a1:  0x00000000  r1/a2:  0x00000080  r2/a3:  0xf44f4b02
[00:00:06.973,000] <err> os: r3/a4:  0x00000006 r12/ip:  0x00000040 r14/lr:  0x00000000
[00:00:06.973,000] <err> os:  xpsr:  0x81000000
[00:00:06.973,000] <err> os: Faulting instruction address (r15/pc): 0x6004adde
[00:00:06.973,000] <err> os: >>> ZEPHYR FATAL ERROR 31: Unknown error on CPU 0
[00:00:06.973,000] <err> os: Current thread: 0x80002b40 (usb_mcux)
[00:00:07.024,000] <err> os: Halting system
```
```
$ addr2line -fe build_dfu/zephyr/zephyr.elf 0x6004adde
USB_DeviceEhciTransfer
/home/jepod/repos/zephyrproject/modules/hal/nxp/mcux/middleware/mcux-sdk-middleware-usb/device/usb_device_ehci.c:1050
```

```
*** Booting Zephyr OS build zephyr-v3.2.0-4025-g0549333eb0d1 ***
[00:00:00.000,000] <inf> main: This device supports USB DFU class.

[00:00:08.084,000] <err> os: ***** USAGE FAULT *****
[00:00:08.084,000] <err> os:   Attempt to execute undefined instruction
[00:00:08.084,000] <err> os: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0x00000010
[00:00:08.084,000] <err> os: r3/a4:  0x00000000 r12/ip:  0xaaaaaaaa r14/lr:  0x60050e2b
[00:00:08.084,000] <err> os:  xpsr:  0x41000000
[00:00:08.084,000] <err> os: Faulting instruction address (r15/pc): 0x60050e16
[00:00:08.084,000] <err> os: >>> ZEPHYR FATAL ERROR 36: Unknown error on CPU 0
[00:00:08.084,000] <err> os: Current thread: 0x800033e0 (idle)
[00:00:08.137,000] <err> os: Halting system
```
```
$ addr2line -fe build_dfu/zephyr/zephyr.elf 0x60050e16
arch_irq_lock
/home/jepod/repos/zephyrproject/zephyr/include/zephyr/arch/arm/aarch32/asm_inline_gcc.h:55
```

danieldegrasse commented:
@jeppenodgaard I'm building the following applications:
`bootloader/mcuboot/boot/zephyr`

<table>
<tr>
	<td> Memory Region
	<td> Used size
<tr>
	<td> FLASH
	<td> 56174 B
<tr>
	<td> RAM
	<td> 24 KB
<tr>
	<td> ITCM
	<td> 8120 B
</table>

`zephyr/samples/subsys/usb/dfu`

<table>
<tr>
	<td> Memory Region
	<td> Used size
<tr>
	<td> FLASH
	<td> 57816 B
<tr>
	<td> RAM
	<td> 20280 KB
<tr>
	<td> ITCM
	<td> 6840 B
</table>

Based off your image sizes it appears we are using different DFU samples. Which sample are you building?

jeppenodgaard commented:
I added these flags to the USB DFU sample: `-DCONFIG_SHELL=y -DCONFIG_FLASH_SHELL=y -DCONFIG_FLASH_MAP_SHELL=y` without them I have the same numbers as you.

Without the additional flags it crashed on my 2nd and 3rd dfu-util transfer attempt.

I used this command:
```
BOARD=mimxrt1050_evk && west build -b $BOARD -d build_bl bootloader/mcuboot/boot/zephyr && west build -b $BOARD zephyr/samples/subsys/usb/dfu -d build_dfu -- -DCONFIG_MCUBOOT_SIGNATURE_KEY_FILE=\"bootloader/mcuboot/root-rsa-2048.pem\" && west build -b $BOARD zephyr/samples/hello_world -d build_hello_world -- -DCONFIG_BOOTLOADER_MCUBOOT=y '-DCONFIG_MCUBOOT_SIGNATURE_KEY_FILE="bootloader/mcuboot/root-rsa-2048.pem"' && west flash -d build_bl && west flash -d build_dfu && sleep 3 && sudo dfu-util --alt 1 --download build_hello_world/zephyr/zephyr.signed.bin
```

jeppenodgaard commented:
@danieldegrasse I do not know if it was clear in my previous comment, but I should be building the same images as you now. However I still have the same issues.

Have you tried USB transfer and power cycle twice several times?

danieldegrasse commented:
Hi @jeppenodgaard, sorry for the delay getting back to you- I have tested this fix with multiple downloads and I can reproduce the issue you are encountering. I'll investigate further and report back.

jeppenodgaard commented:
> Hi @jeppenodgaard, sorry for the delay getting back to you- I have tested this fix with multiple downloads and I can reproduce the issue you are encountering. I'll investigate further and report back.

No problem. Thank you, I really appreciate your help on this!

danieldegrasse commented:
Hi @jeppenodgaard, I just wanted to give you an update on this- I unfortunately have not found the root cause of this issue, but I do believe it is not directly related to the hyperflash write bug. I have reproduced a similar issue on the RT1060 EVK, which has a different flash driver. I've opened a bug tracking that issue on the RT1050 and RT1060 here: #54731

As far as this PR, I am ok to wait on the fix if you'd prefer to test with the DFU sample. However, I do think that the flash driver on the RT1050 is functional with your changes (although the USB may not be)

jeppenodgaard commented:
Thank you for the update @danieldegrasse! I have tried to test with MCU mgr serial but that also results in a crash when I try to transfer. The exception is not consistent (similar to dfu-util transfer). Maybe you can use this information to find the root cause.

I test with this script:
```
#!/bin/bash

BOARD=mimxrt1050_evk

west build -b $BOARD -d build_mcuboot bootloader/mcuboot/boot/zephyr
west build -b $BOARD zephyr/samples/subsys/mgmt/mcumgr/smp_svr -- \
-DOVERLAY_CONFIG='overlay-serial.conf' -DCONFIG_BOOTLOADER_MCUBOOT=y \
'-DCONFIG_MCUBOOT_SIGNATURE_KEY_FILE="bootloader/mcuboot/root-rsa-2048.pem"'

west flash -d build_mcuboot
west flash

mcumgr --conntype serial --connstring "/dev/ttyUSB0,baud=115200" image list
mcumgr --conntype serial --connstring "/dev/ttyUSB0,baud=115200" image upload build/zephyr/zephyr.signed.bin
```

And I added these lines to `mimxrt1050_evk.dts `:
```
	chosen {
		zephyr,uart-mcumgr=&lpuart3;
```
```
&lpuart3 {
	status = "okay";
	current-speed = <115200>;
	pinctrl-0 = <&pinmux_lpuart3>;
	pinctrl-1 = <&pinmux_lpuart3_sleep>;
	pinctrl-names = "default", "sleep";
};
```

danieldegrasse commented:
@jeppenodgaard Are you testing with this commit in NXP's HAL: https://github.com/nxp-zephyr/hal_nxp/commit/8f5c9b7c9821da65b44860708bea8d0b3633c2b3? Without that change I reliably encounter issues with MCUMgr, but when it is present I have tested running up to 5 transfers in a row without issue (and verified the upgrade process works after those transfers)

jeppenodgaard commented:
@danieldegrasse I am not sure. I'll check tomorrow.

jeppenodgaard commented:
@danieldegrasse I should be building with the relocation changes:
[mcumgr-serial-test.md](https://github.com/zephyrproject-rtos/zephyr/files/10753450/mcumgr-serial-test.md)

danieldegrasse commented:
@jeppenodgaard I was able to reproduce your issue. I believe I was testing with my local branch before, where I'd applied a few potential fixes for the USB issue you were seeing (while debugging it). Take a look at this commit, the changes here seem to solve the issue for me: https://github.com/nxp-zephyr/zephyr/commit/36b0b7f06e2ff8f01b5f38c2bd5d5f6c8c4e352e

jeppenodgaard commented:
@danieldegrasse I tried testing again with your commits both serial and USB. I based them the branch on top of the USB fix PR.
It feels more stable but it still crashes sometimes.

I've created [this branch](https://github.com/jeppenodgaard/zephyr/tree/bugfix/mimxrt1050_hyperflash_write_2) for testing.
I ran `for i in $(seq 100) : ; do ( echo $i; ./dfu-usb.sh asd ) || break; done` to test transfer stability.

I also came across [this issue](https://github.com/zephyrproject-rtos/zephyr/issues/44647) where you suggest using progressive erase, which I added as a build flag in the scripts.

danieldegrasse commented:
@jeppenodgaard I have been testing with your branch, and thus far I have been unable to reproduce a failure on my RT1050 EVK.

I'm using:
- Zephyr SHA 8a97ebee95362a8428740901c95ef4c0196f5cd3 (https://github.com/jeppenodgaard/zephyr/commit/8a97ebee95362a8428740901c95ef4c0196f5cd3)
- NXP HAL SHA: 8f5c9b7c9821da65b44860708bea8d0b3633c2b3 (https://github.com/nxp-zephyr/hal_nxp/commit/8f5c9b7c9821da65b44860708bea8d0b3633c2b3

I ran
```
$ ./dfu-usb.sh # to flash sample to RT1050
$ for i in $(seq 100) : ; do ( echo $i; ./dfu-usb.sh asd ) || break; done # to test stability
```

I've also tried the mcumgr test, using the following commands (I of course had to modify `dfu-serial.sh` to use the correct TTY device for my setup):
```
$ ./dfu-serial.sh # to flash sample to RT1050
$ for i in $(seq 100) : ; do ( echo $i; ./dfu-serial.sh asd ) || break; done
```

I've run this test a few times now and I've yet to see an error. What output do you see when one occurs?

jeppenodgaard commented:
@danieldegrasse I see various exceptions. Strange that it works on your end, maybe your hardware performs better :thinking: 

I was able to avoid the crash by out-commenting the line in `FLEXSPI_UpdateDllValue` that sets `DLLCR`.
I've used some code from `mflash_drv_page_program_internal` that changes the `DIV` and `DLL` in a more generic way and I am not able to make it crash anymore. The bootloader is also able to use the image.
I probably tested the bootloader 10 times and the transfer 50.

I've added the wip changes to the branch in a new commit.

I hope this works for you as well :crossed_fingers:

jeppenodgaard commented:
@danieldegrasse I anticipate this will work for you as well. I will clean up the commits if that is ok with you or do you want to test it first?

danieldegrasse commented:
@jeppenodgaard the fixes still work for me with your updates. Are you seeing a different value returned from `flexspi_clock_get_freq` versus `CLOCK_GetClockRootFreq(kCLOCK_FlexspiClkRoot);`? On my board, I do not. I just want to make sure that these fixes aren't masking an underlying issue, by introducing a delay where there wasn't one before. One thing the MCUXpresso SDK examples do that our driver does not is disable the SCB cache before performing flexSPI accesses: https://github.com/nxp-mcuxpresso/mcux-sdk-examples/blob/main/evkbimxrt1050/driver_examples/flexspi/hyper_flash/polling_transfer/flexspi_hyper_flash_ops.c#L31

Does calling `SCB_DisableDCache` and `SCB_DisableICache` before erase or writing to the flash (and then calling the associated `SCB_Enable*` functions afterwards) fix your issue, without the more generic clock calculation present in your latest commit?

jeppenodgaard commented:
I added a print before the `return` in `memc_flexspi_update_clock`.

On `bugfix/mimxrt1050_hyperflash_write`
```
div 3, flexspi_clock_get_freq() 65454543, CLOCK_GetClockRootFreq(kCLOCK_FlexspiClkRoot) 65454543
div 0, flexspi_clock_get_freq() 261818172, CLOCK_GetClockRootFreq(kCLOCK_FlexspiClkRoot) 261818172
```

On `bugfix/mimxrt1050_hyperflash_write_2`
```
div 2, flexspi_clock_get_freq() 87272724, CLOCK_GetClockRootFreq(kCLOCK_FlexspiClkRoot) 87272724
div 0, flexspi_clock_get_freq() 261818172, CLOCK_GetClockRootFreq(kCLOCK_FlexspiClkRoot) 261818172
```

The get clock functions return the same value. However the divider is different.

Adding `SCB_DisableDCache(); SCB_DisableICache();` after `irq_lock()` and `SCB_EnableDCache(); SCB_EnableICache();` before `irq_unlock()` does not seem to change anything.

Setting the divider to `2` instead of `3` in `bugfix/mimxrt1050_hyperflash_write` makes it run stable. This does not make much sense to me since the clock then runs faster while writing. Setting the divider to 5 which makes `flexspi_clock_get_freq()` return a value < 50 MHz, which as far as I can tell is the value we are looking for, also make the system crash.

danieldegrasse commented:
> Setting the divider to 2 instead of 3 in bugfix/mimxrt1050_hyperflash_write makes it run stable. This does not make much sense to me since the clock then runs faster while writing. Setting the divider to 5 which makes flexspi_clock_get_freq() return a value < 50 MHz, which as far as I can tell is the value we are looking for, also make the system crash.

This is strange to me as well. I'm also surprised to see your clock dividers differ. Unless I made a mistake debugging this, I did not see any difference on my board. Are you using different clock settings for SOC initialization, or using a custom board? If you are, I'm wondering if there is anything that might make your hardware behave differently from the EVK.

jeppenodgaard commented:
@danieldegrasse I am using the dfu-usb.sh which is using the mimxrt1050_evk. So nothing custom.

danieldegrasse commented:
> @danieldegrasse I am using the dfu-usb.sh which is using the mimxrt1050_evk. So nothing custom.

That is strange. I think one thing to try is to move the USB DFU sample to run from RAM, to eliminate any potential read-while-write issues that could stem from executing from flash while writing to it.

I've created a branch here where the `dfu-usb.sh` script is updated to build the USB DFU sample (and hello world sample) for RAM loading with MCUBoot: https://github.com/nxp-zephyr/zephyr/tree/bugfix/mimxrt1050_hyperflash_write_ramload. There is still some issue (likely with a missing Kconfig) that makes booting the hello world sample unreliable, but the USB DFU sample runs and transfers data reliably.

Could you see if the USB transfers are stable for you using this branch? Note that once the `hello_world` sample is written to flash, MCUBoot will always boot it at reset (since MCUBoot only uses the image version to determine which slot to load to RAM when using RAMLOAD mode), so to switch to the USB_DFU sample again you'll need to erase the `hello_world` sample from flash (or increase the image version of the USB_DFU sample to 0.0.2)

jeppenodgaard commented:
@danieldegrasse Good idea to try running it from flash and thank you for creating the script to do so. I do not have access to the hardware atm. I will try it in the morning :)

jeppenodgaard commented:
@danieldegrasse I am not able to build `zephyr/samples/subsys/usb/dfu`:
```
/home/jepod/zephyr-sdk-0.15.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/../../../../arm-zephyr-eabi/bin/ld.bfd:zephyr/linker_zephyr_pre0.cmd:56 cannot move location counter backwards (from 0000000020201440 to 0000000020200200)
```
I can increase `CONFIG_ROM_START_OFFSET` but does it work for you with the 0x200 value?

danieldegrasse commented:
> @danieldegrasse I am not able to build `zephyr/samples/subsys/usb/dfu`:
> 
> ```
> /home/jepod/zephyr-sdk-0.15.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/../../../../arm-zephyr-eabi/bin/ld.bfd:zephyr/linker_zephyr_pre0.cmd:56 cannot move location counter backwards (from 0000000020201440 to 0000000020200200)
> ```
> 
> I can increase `CONFIG_ROM_START_OFFSET` but does it work for you with the 0x200 value?

Sorry about that, I should have verified a clean build worked! I've updated the script, and a clean build is working for me. It seems that the RT ROM header was still being included in the image (which is not required)

jeppenodgaard commented:
> Sorry about that, I should have verified a clean build worked! I've updated the script, and a clean build is working for me. It seems that the RT ROM header was still being included in the image (which is not required)
>
I removed commit e9701493a5 and tested executing from RAM - it no longer crashes. So I guess it is a read while write issue.
I am not sure how to debug which code should be and is not relocated.

jeppenodgaard commented:
@danieldegrasse Do you also think this is narrowed down to a read-while-write issue?

Commit:
drivers: flash: check xip before configure device in hyperflash init

Check memc_flexspi_is_running_xip before calling
memc_flexspi_set_device_config in hyperflash init.
This aligns with the nor flash driver.

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

jeppenodgaard commented:
@danieldegrasse I have attempted to find any variables or functions that are not located in RAM or ITCM in the write and erase function. Everything except memcpy usage looks good to me and fixing that does not stop it from crashing.

I have however been able to narrow the crash down a bit more. Earlier I mentioned that I was not able to make it crash if I outcommented the code that related to setting the DLLxCR register. I found that crash does not occur if I constantly set configured the DLLxCR register for slow clock.

I then outcommented all the code between `irq_lock` and `irq_unlock` in the write function except for `(void)memc_flexspi_update_clock(&data->controller, &data->config, data->port, MEMC_FLEXSPI_CLOCK_166M);` which would still make the system crash.

Based on those findings I am not convinced it is a read-while-write issue anymore.

I noticed that `flash_flexspi_nor_init` conditionally executes `memc_flexspi_set_device_config`. If I use that logic in the hyperflash driver there is no longer an issue when changing the DLLxCR register and I am not able to make it crash so far.

@pdgendt Do you know why that was added to the nor driver?

I added a commit with that change. I also create a new branch https://github.com/jeppenodgaard/zephyr/tree/bugfix/mimxrt1050_hyperflash_write_3. There's a build.sh and test.sh script that I've used. The test.sh makes it a lot easier to test the stability because it resets the MCU.

pdgendt commented:
> I noticed that `flash_flexspi_nor_init` conditionally executes `memc_flexspi_set_device_config`. If I use that logic in the hyperflash driver there is no longer an issue when changing the DLLxCR register and I am not able to make it crash so far.
> 
> @pdgendt Do you know why that was added to the nor driver?

The configuration was skipped for devices running in XiP, where ROM code would already have set the configuration before starting the application, or even a bootloader.

jeppenodgaard commented:
@danieldegrasse how does the new change look to you?

de-nordic commented on samples/drivers/flash_shell/src/main.c:
```diff
@@ -160,30 +161,35 @@ static int parse_u8(const char *str, uint8_t *result)
-	if (ret) {
-		goto err_read;
+		dump_buffer(sh, buf, read_len);
+		if (cmp_buf != NULL) {
+			cmp_ok &= memcmp(cmp_buf, buf, read_len) == 0;
+			cmp_buf += read_len;
+		}
```
```
if (cmp_ok && cmp_buf !=NULL) {
  cmp_ok = ...
}
```
if `cmp_ok` gets false, we are no longer interested in comparing anything, we already know buffer is bad.
Comparing every buffer only makes sense if you are going to notify failure separately for each buffer.

jeppenodgaard replied:
That makes sense. I have tried to add you suggestion in a [separate commit](https://github.com/zephyrproject-rtos/zephyr/pull/53857/commits/f903901e1eaa90f570d4db6fd14ccb812798d61a). Here is a sample output:
![image](https://user-images.githubusercontent.com/4233729/233992232-bd2bb312-dfee-485d-bef6-83e94bc73c52.png)

jeppenodgaard replied:
Note that I modified the write_pattern locally to make the read back fail.

de-nordic replied:
> Note that I modified the write_pattern locally to make the read back fail.

Looks nice. I would also print the "Write verification error, unexpected values marked red" or something like that, because it is not clear why there are red

jeppenodgaard replied:
That's a good idea. I've changed the commit a bit.

de-nordic commented on samples/drivers/flash_shell/src/main.c:
```diff
@@ -160,30 +161,35 @@ static int parse_u8(const char *str, uint8_t *result)
-		ret = flash_read(flash_device, offset, buf, sizeof(buf));
-		if (ret) {
-			goto err_read;
+	size_t read_len;
```
You can move this to `do {` , this is only scope where it is used.

danieldegrasse reviewed (commented):
> @danieldegrasse how does the new change look to you?

Specifically regarding the changes for the hyperflash driver, I am generally OK with the approach here. Please see my comment about the `CLOCK_SetDiv` function though

jeppenodgaard commented on soc/arm/nxp_imx/rt/flexspi_rt10xx.c:
```diff
@@ -0,0 +1,29 @@
+{
+	CLOCK_DisableClock(kCLOCK_FlexSpi);
+
+	CLOCK_SetDiv(kCLOCK_FlexspiDiv, value);
```
I used the commit that relocates the FSL clock driver:
https://github.com/nxp-zephyr/hal_nxp/commit/8f5c9b7c9821da65b44860708bea8d0b3633c2b3
Would you rather that I implement a Zephyr version for non-inline functions?

jeppenodgaard replied:
@danieldegrasse I am not sure if I should tag you when replying so I will do that now :smile:

de-nordic reviewed (commented):
@jeppenodgaard Can you take shell related commits to separate PR.
They are unrelated to fix and do not have to wait for the rest of this PR to go through.
Thanks.

jeppenodgaard commented:
> @jeppenodgaard Can you take shell related commits to separate PR. They are unrelated to fix and do not have to wait for the rest of this PR to go through. Thanks.

Created https://github.com/zephyrproject-rtos/zephyr/pull/57233.

pdgendt commented on drivers/flash/flash_mcux_flexspi_hyperflash.c:
```diff
@@ -418,6 +418,10 @@ static int flash_flexspi_hyperflash_write(const struct device *dev, off_t offset
 		return -EINVAL;
 	}
 
+#ifdef CONFIG_HAS_MCUX_CACHE
+	SCB_DisableICache();
+#endif
```
There is a cache API in zephyr, shouldn't we use that? Or isn't that available for i.MX RT series?
```suggestion
	sys_cache_instr_disable();
```

danieldegrasse committed:
drivers: flash: hyperflash driver no longer stores controller data in ROM

Store controller reference in RAM, to avoid flash access in critical
section.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

Commit:
drivers: flash: fix memcpy and invalidate dcache in hyperflash

Replace memcpy usage in write since it is not linked to non-external
flash which might cause read-while-write issues.
Move the function that invalidated the data cache inside the critical
section to avoid context switch before it is invalidated.

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

jeppenodgaard commented:
I still had rare instability issues which seemed to be improved by storing the controller in RAM (I used your commit @danieldegrasse and removed wip from the commit message), moving the DCache invalidate into the critical section and remove memcpy usage.
I also removed the ICache calls since they did not seem improve stability.

jeppenodgaard commented:
It is still not 100% stable.. I am trying to find the issue.

jeppenodgaard commented:
> It is still not 100% stable.. I am trying to find the issue.

It turned out that it was not working on the v3.3-branch but stable on the main branch.

jeppenodgaard commented:
ping @danieldegrasse

danieldegrasse commented on drivers/flash/flash_mcux_flexspi_hyperflash.c:
```diff
@@ -427,6 +427,9 @@ static int flash_flexspi_hyperflash_write(const struct device *dev, off_t offset
 		key = irq_lock();
 	}
 
+	(void)memc_flexspi_update_clock(data->controller, &data->config,
```
Minor comment- the description for your commit "drivers: flash: fix hyperflash write operations" currently mentions disabling the instruction cache, but I don't see any calls to do that in the PR. Could you update the commit message?

jeppenodgaard replied:
Thanks! I fixed it.

jeppenodgaard commented:
@pdgendt can I persuade you to do another review?
